asdf/util.py

```diff
@@ -2,13 +2,27 @@
 import math
 import struct
 import types
+import importlib.util
 
-from urllib.parse import urljoin
 from urllib.request import pathname2url
-from urllib import parse as urlparse
 
 import numpy as np
 
+# We're importing our own copy of urllib.parse because
+# we need to patch it to support asdf:// URIs, but it'd
+# be irresponsible to do this for all users of a
+# standard library.
+urllib_parse_spec = importlib.util.find_spec('urllib.parse')
+patched_urllib_parse = importlib.util.module_from_spec(urllib_parse_spec)
+urllib_parse_spec.loader.exec_module(patched_urllib_parse)
+del urllib_parse_spec
```

Comment:
What the advantage of doing it this way rather than

```
import urllib.parse as patched_urllib_parse
```

It's still local to this module's namespace, right?

Reply:
Unfortunately while that aliases the module reference to a new name, it still points to the same module object.  Import caches the module after loading it the first time so that other imports don't have to read and parse the module's code again.

```python
# module1.py
import urllib.parse as module1_alias

# module2.py
import urllib.parse as module2_alias

# Fire up a console:
In [1]: import module1

In [2]: import module2

In [3]: module1.module1_alias is module2.module2_alias
Out[3]: True
```

Reply:
Ah, I see.  I didn't realize there was caching going on.  This bypasses that. #themoreyouknow